src/cover_crypt_core.rs

```diff
@@ -466,19 +479,23 @@ where
         + Mul<&'b KeyPair::PrivateKey, Output = KeyPair::PrivateKey>
         + Div<&'b KeyPair::PrivateKey, Output = KeyPair::PrivateKey>,
 {
-    let mut res = Vec::with_capacity(sk_j.x.len() * encapsulation.E.len());
     let precomp = &(&encapsulation.C * &sk_j.a) + &(&encapsulation.D * &sk_j.b);
-    for E_i in &encapsulation.E {
+    for (tag_i, E_i) in &encapsulation.E {
         for x_k in sk_j.x.values() {
-            let K_k = &precomp * x_k;
-            let mut K = hkdf_256::<SYM_KEY_LENGTH>(&K_k.to_bytes(), KEY_GEN_INFO.as_bytes())?;
-            for (e_1, e_2) in K.iter_mut().zip(E_i.as_bytes()) {
-                *e_1 ^= e_2;
+            let mut hasher = Sha3_512::new();
+            hasher.update((&precomp * x_k).to_bytes());
+            hasher.update(KEY_GEN_INFO);
+            let mut bytes = hasher.finalize();
+            if tag_i.as_bytes() != &bytes[..SYM_KEY_LENGTH] {
+                continue;
+            }
+            for (b1, b2) in bytes[SYM_KEY_LENGTH..].iter_mut().zip(E_i.as_bytes()) {
+                *b1 ^= b2;
             }
-            res.push(SymmetricKey::from_bytes(K));
+            return Ok(SymmetricKey::try_from_bytes(&bytes[SYM_KEY_LENGTH..])?);
```

Comment:
can you add a small comment about why you choose to either continue or return early here, just to make it a bit more explicit